# Laboratorium 2


In [2]:
import numpy as np

## Zadanie 1.6[MG] 

Napisz funkcje, ktora wyznacza modę z danego obiektu iterowalnego przechowujacego
obiekty niezmienne jako elementy. Jesli moda nie jest jednoznaczna, zwroć:
- wartość dowolną;
- wartość losowo wybraną.

In [4]:
from collections import Counter
import random

def moda_dowolna(iterable):
    """Zwraca modę (najczęstszy element). Przy remisie — dowolny z najczęstszych."""
    licznik = Counter(iterable)
    return licznik.most_common(1)[0][0]

def moda_losowa(iterable):
    """Zwraca modę. Przy remisie — losowo wybrany spośród najczęstszych."""
    licznik = Counter(iterable)
    max_count = licznik.most_common(1)[0][1]
    najczestsze = [k for k, v in licznik.items() if v == max_count]
    print(najczestsze)
    return random.choice(najczestsze)

In [5]:
dane = [1, 2, 2, 3, 3, 4]

print(Counter(dane))

Counter({2: 2, 3: 2, 1: 1, 4: 1})


In [6]:
Counter(dane).most_common(2)

[(2, 2), (3, 2)]

In [7]:
Counter(dane).most_common(2)[1][1]

2

In [8]:

print(moda_dowolna(dane))   # zawsze 2 (first encountered by Counter)
print(moda_losowa(dane))    # 2 lub 3 losowo

2
[2, 3]
3


## Zadanie 1.8 [MG] 

Dla danej 1-wymiarowej tablicy o elementach liczbowych (np. klasy ndarray) x wykonaj
następujące operacje.

a) Wypisz na konsolę wszystkie wartości ze zbioru [−2,−1] ∪ [1, 2].  
b) Wypisz na konsolę liczbę oraz frakcję wszystkich wartości nieujemnych.  
c) Wyznacz średnią arytmetyczną wartości bezwzględnych elementów.  
d) Wyznacz wartość najbliższą i najdalszą od 0 (zachowując jej znak).  
e) Wyznacz wartość najbliższą i najdalszą od 2 (zachowując jej znak).  
f) Wypisz na konsolę wektor powstały w wyniku przekształcenia liniowego wartości z x na przedział [0, 1] (najmniejsza wartość staje się równa 0, a największa 1).  
g) Utwórz listę skłądającą się z dwóch wektorów: pierwszy zawiera wszystkie wartości nieujemne z x, a drugi wszystkie wartości ujemne.  
h) Utwórz wektor napisów y o długości takiej samej, jaką ma x, dla którego yi przyjmuje wartość "nieujemna", jeśli xi jest nieujemne oraz "ujemna" w przeciwnym przypadku.  
i) Utwórz wektor napisów y o długości takiej samej, jaką ma x, dla którego yi przyjmuje wartość "mały", jeśli $x_i < −1$, "średni", dla $|x_i| ≤ 1$ oraz "duży" w przeciwnym przypadku.  
j) Utwórz wektor liczbowy y o długości takiej samej, jaką ma x, dla którego yi przyjmuje wartość k + 1/2 wtedy i tylko wtedy, gdy $x_i ∈ [k, k + 1)$, gdzie k ∈ Z (prosty histogram).

In [10]:
# ustawiamy ziarno generatora liczb pseudolosowych
# korzystamy z modułu random pakietu numpy
np.random.seed(6)

# tworzymy tablicę jednowymiarwą (wektor)
# złożony z wartości wylosowanych z rozkładu standardowego normalnego
x = np.round(np.random.normal(size=20), 2)
# np.round() zaokrągla wartość liczbową - u nas do 2 miejsc po przecinku
x

array([-0.31,  0.73,  0.22, -0.9 , -2.49,  0.91,  1.13, -1.51,  1.64,
       -0.43,  2.63,  0.6 , -0.34,  1.24,  0.11,  0.13,  0.08, -0.16,
        0.63,  0.81])

**a) Wypisz na konsolę wszystkie wartości ze zbioru [−2,−1] ∪ [1, 2].**

In [12]:
# to wyrażenie da nam tablicę logiczną
((x>=-2) & (x<=-1)) | ((x >= 1) & (x <= 2))

array([False, False, False, False, False, False,  True,  True,  True,
       False, False, False, False,  True, False, False, False, False,
       False, False])

In [13]:
# element odpowiadający True zostaje wybrany
# element odpowiadający False nie

x[((x>=-2) & (x<=-1)) | ((x >= 1) & (x <= 2))]

array([ 1.13, -1.51,  1.64,  1.24])

**(b) Wypisz na konsolę liczbę oraz frakcję wszystkich wartości nieujemnych.**

In [15]:
x>=0

array([False,  True,  True, False, False,  True,  True, False,  True,
       False,  True,  True, False,  True,  True,  True,  True, False,
        True,  True])

In [16]:
# liczba wartości nieujemnych:
np.sum(x>=0)

np.int64(13)

In [17]:
# frakcja - proporcja:

np.sum(x>=0)/len(x) # zamiast len() możemy użyć x.shape[0]
                    # len(y) == y.shape[0] 

np.float64(0.65)

In [18]:
# lub z wykorzystanim funkcji np.mean()
np.mean(x>=0)

np.float64(0.65)

In [19]:
# możemy skorzystać z metody .mean() dla tablic:
(x >= 0).mean()

np.float64(0.65)

In [20]:
np.mean(x>=0)

np.float64(0.65)

**(c) Wyznacz średnią arytmetyczną wartości bezwzględnych elementów.**

In [22]:
# wystarczy skorzystać z funkcji np.mean() oraz np.abs()
# przy czym funkcja np.abs() jest zwektoryzowana
np.mean(np.abs(x))

np.float64(0.85)

In [23]:
# lub
np.abs(x).mean()

np.float64(0.85)

**(d) Wyznacz wartość najbliższą i najdalszą od 0 (zachowując jej znak).**

In [25]:
np.abs(x)

array([0.31, 0.73, 0.22, 0.9 , 2.49, 0.91, 1.13, 1.51, 1.64, 0.43, 2.63,
       0.6 , 0.34, 1.24, 0.11, 0.13, 0.08, 0.16, 0.63, 0.81])

In [26]:
np.argmin(np.abs(x))

np.int64(16)

In [27]:
# wartość nambliższa od 0

x[np.argmin(np.abs(x))]
# funkcja np.argmin() zwraca indeks namniejszego elementu
# (pierwszego wystąpienia najmniejszego elementu:
# np. np.argmin(np.array([1, 4, 1, 5, 6, 1, 6])) == 0

np.float64(0.08)

In [28]:
x[np.argmax(np.abs(x))]
# funkcja np.argmax() analogicznie jak wyżej
# tylko największego elementu
# np.
# np.argmax(np.array([1, 4, 1, 5, 6, 1, 6])) == 4

# odpowiedniki z R: which.min, which.max

np.float64(2.63)

**(e) Wyznacz wartość najbliższą i najdalszą od 2 (zachowując jej znak).**

In [30]:
np.abs(x-2) # ,,odległość'' poszczególnych elementów od 2

array([2.31, 1.27, 1.78, 2.9 , 4.49, 1.09, 0.87, 3.51, 0.36, 2.43, 0.63,
       1.4 , 2.34, 0.76, 1.89, 1.87, 1.92, 2.16, 1.37, 1.19])

In [31]:
y = np.abs(x-2)
x[np.argmin(y)], x[np.argmax(y)]

(np.float64(1.64), np.float64(-2.49))

**(f) Wypisz na konsolę wektor powstały w wyniku przekształcenia liniowego wartości z 
 na przedział (najmniejsza wartość staje się równa 0, a największa 1).**

In [33]:
# od każdego elementu odejmujemy minimum z x -> przesuwamy do 0
# każdy element dzielimy przez ,,nowe'' maksimum (x.max() - x.min())
(x - x.min())/(x.max() - x.min())

array([0.42578125, 0.62890625, 0.52929688, 0.31054688, 0.        ,
       0.6640625 , 0.70703125, 0.19140625, 0.80664062, 0.40234375,
       1.        , 0.60351562, 0.41992188, 0.72851563, 0.5078125 ,
       0.51171875, 0.50195312, 0.45507812, 0.609375  , 0.64453125])

**(g) utwórz listę składającą się z dwóch wektorów: pierwszy zawiera wszystkie watości nieujemne z x, drugi wszystkie wartości ujemne.**

In [35]:
# tworzymy liste składającą się z 2 tablic numpy.ndarray
# pierwsza ma wartosci ujemne z x
# druga nieujemne
[ x[x < 0], x[x >= 0] ]

[array([-0.31, -0.9 , -2.49, -1.51, -0.43, -0.34, -0.16]),
 array([0.73, 0.22, 0.91, 1.13, 1.64, 2.63, 0.6 , 1.24, 0.11, 0.13, 0.08,
        0.63, 0.81])]

**(h) Utwórz wektor napisów o długości takiej samej, jaką ma, dla którego 
 przyjmuje wartość "nieujemna", jeśli jest nieujemne oraz "ujemna" w przeciwnym przypadku.**

In [37]:
# w rozwiązaniu wykorzystamy funkcję np.where (zajrzyj do jej dokumentacji)
# przykład zastosowania:
np.where(x < 0) # dostajemy indeksy elementów < 0

(array([ 0,  3,  4,  7,  9, 12, 17]),)

In [38]:
# lub możemy wykorzystać ją
# dostaniemy w wyniku tablicę
# tam gdzie warunek jest spełniony wrzucana jest pierwsza wartość (ujemna)
# tam gdzie nie jest spełniony - druga (nieujemna)
np.where(x < 0, "ujemna", "nieujemna")

array(['ujemna', 'nieujemna', 'nieujemna', 'ujemna', 'ujemna',
       'nieujemna', 'nieujemna', 'ujemna', 'nieujemna', 'ujemna',
       'nieujemna', 'nieujemna', 'ujemna', 'nieujemna', 'nieujemna',
       'nieujemna', 'nieujemna', 'ujemna', 'nieujemna', 'nieujemna'],
      dtype='<U9')

**(i) Utwórz wektor napisów 
 o długości takiej samej, jaką ma 
, dla którego 
 przyjmuje wartość "mały", jeśli 
, "średni", dla 
 oraz "duży" w przeciwnym przypadku.**

In [40]:
# Podobnie tylko zagnieżdżamy:
np.where(x <- 1, "maly", np.where(x > 1, "duzy", "sredni"))

array(['sredni', 'sredni', 'sredni', 'sredni', 'maly', 'sredni', 'duzy',
       'maly', 'duzy', 'sredni', 'duzy', 'sredni', 'sredni', 'duzy',
       'sredni', 'sredni', 'sredni', 'sredni', 'sredni', 'sredni'],
      dtype='<U6')

**(j) Utwórz wektor liczbowy $\mathtt{y}$ o długości takiej samej, jaką
ma $\mathtt{x}$, dla którego $y_i$ przyjmuje wartość
$k+1/2$ wtedy i tylko wtedy, gdy $x_i\in[k,k+1)$, gdzie $k\in\mathbb{Z}$
(prosty histogram).**

In [42]:
# np.floor() funkcja zwektoryzowana, zwraca podłogę z poszczególnych elementów wektora
np.floor(x) + 0.5

array([-0.5,  0.5,  0.5, -0.5, -2.5,  0.5,  1.5, -1.5,  1.5, -0.5,  2.5,
        0.5, -0.5,  1.5,  0.5,  0.5,  0.5, -0.5,  0.5,  0.5])

## Zadanie 1.9[MG] 

Napisz funkcję, która dokonuje standaryzacji wartości wektora numerycznego, tj., dokonuje przekształcenia jego wartości tak by średnia wynikowego wektora była równa 0, a odchylenie standardowe 1 (or. Z-scoring).

In [44]:
def standardize(x):
    """
    Funkcja dokonuje standaryzacji wektora x
    Parameters
    ----------
    x : tablica numpy.ndarray jednowymiarowa  
        (wektor)

    Returns
    -------
    Tablica jednowymiarowa będąca wystandaryzowaną wersją x
    
    """
    
    assert x.ndim == 1 # czy wymiar odpowiedni
    # w przypadku tablic numpy.ndarray zachodzi uzgadnianie kształtów -
    # przez przypadek możemy więc (bez błędu) policzyć coś co nie jest 
    # poprawnym (oczekiwanym) przez nas wynikiem.
    # Przed taką sytuacją się zabezpieczamy prz użyciu assert.
    #
    # Korzystamy z wektoryzacji:
    # 1.) od każdego elementu odejmujemy średnia: x - x.mean()
    # 2.) każdy tak zmieniony element dzielimy przez odchylenie: 
    #     (...) / x.std(ddof = 1)
    # Uwaga:
    # wariancja = (\sum_{i=1}^{n} (x[i] - srednia_x)**2) / (n - ddof)
    # dla ddof = 1 mamy najlepsze własności statystyczne
    # ddof=1 (jak w kodzie) — estymator nieobciążony wariancji, dzieli przez n-1. Używany gdy x to próbka i chcemy estymować odchylenie standardowe populacji.
    # ddof=0 — dzieli przez n. Używany gdy x to cała populacja albo gdy robimy czysto techniczną normalizację (np. preprocessing w ML, batch normalization).
    
    return (x - x.mean()) / x.std(ddof = 1)

In [45]:
standardize(x)

array([-0.48705389,  0.44066781, -0.01427264, -1.01335755, -2.43170129,
        0.60123502,  0.79748384, -1.557502  ,  1.25242429, -0.5940987 ,
        2.13554398,  0.32470259, -0.51381509,  0.89560825, -0.11239705,
       -0.09455625, -0.13915825, -0.35324788,  0.3514638 ,  0.51203101])

In [46]:
x = np.random.exponential(scale = 10, size = 100)
x

array([12.34884866,  7.78567506,  1.33330469, 31.61722999,  5.16267569,
        2.4456021 , 12.63283579, 51.51180671,  2.95195636, 11.12638493,
        9.13808598, 12.6348273 , 27.7018417 ,  4.33571059,  2.92539313,
        5.14954986, 13.72433734, 12.87610125,  5.21057451, 45.45001977,
        5.98745196,  4.68098939, 12.36594507,  0.86057527,  5.08117492,
       14.73513613, 14.47110681,  3.3387874 ,  2.10155994,  6.37241373,
        4.07676618, 13.27026117,  2.09619312,  4.02126798, 18.59364581,
        9.54574373, 21.65122137, 37.53013594, 18.13444456,  1.99457086,
        9.63409878,  6.4647139 ,  5.1777749 , 13.54344972, 17.48316658,
       11.49578785,  1.49976278, 12.13434281,  0.60639243,  2.14017167,
       25.87521139,  5.19366086,  1.60510985, 11.46607231,  1.78515311,
       10.44764198,  2.89836583,  0.42255836, 31.87440152,  0.68514929,
        7.19368342,  4.17481908, 10.7883854 , 18.60642371,  9.26004146,
        8.84825231,  1.24467492, 12.39428444,  2.28813668,  6.74

In [47]:
standardize(x)

array([ 0.26764826, -0.2130604 , -0.89278698,  2.29748073, -0.48938093,
       -0.77561173,  0.29756496,  4.39327991, -0.72226971,  0.13886752,
       -0.07059034,  0.29777476,  1.88501317, -0.57649778, -0.72506802,
       -0.49076367,  0.41254946,  0.32319182, -0.48433502,  3.75469944,
       -0.40249467, -0.5401243 ,  0.26944929, -0.94258678, -0.49796665,
        0.51903232,  0.49121808, -0.6815189 , -0.81185494, -0.36194077,
       -0.60377634,  0.3647147 , -0.81242031, -0.60962281,  0.92550799,
       -0.02764553,  1.24760905,  2.92037729,  0.87713332, -0.82312573,
       -0.01833774, -0.35221739, -0.4877903 ,  0.39349381,  0.80852427,
        0.17778236, -0.87525141,  0.24505109, -0.96936373, -0.80778738,
        1.69258634, -0.48611679, -0.8641536 ,  0.17465197, -0.8451869 ,
        0.06736517, -0.72791522, -0.98872978,  2.32457253, -0.96106707,
       -0.27542391, -0.59344693,  0.10326087,  0.92685408, -0.05774291,
       -0.10112294, -0.90212371,  0.27243471, -0.79219996, -0.32

In [48]:
def standardize2(x):
    
    import pandas as pd
    from sklearn.preprocessing import MinMaxScaler, StandardScaler
    
    scaler = StandardScaler()
    #scaler = MinMaxScaler()
    output = scaler.fit_transform(pd.DataFrame(x))
    
    # outupt2 = scaler.transform(output)
    # output.mean()
    # output.std()
    return output


In [49]:
standardize2(x)

ModuleNotFoundError: No module named 'sklearn'

## Zadanie 1.11[MG&AC&MB]

Zaimplementuj funkcję, która dla danego wektora $\mathtt{x}$ o długości $d$ oraz nieparzystej liczby naturalnej $k$ wyznaczy średnią ruchomą, $k<d$, tj. taki wektor $(w_1, w_2, \dots, w_{d-k+1})$, że
$$w_i = \sum_{j=1}^{k}\frac{x_{i+j-1}}{k}.$$


In [ ]:
import numpy as np
import matplotlib.pyplot as plt

def srednia_ruchoma(x, k):
    """
    Funkcja wyznacza średnią ruchomą dla danego
    wektra x
    
    Więcej szczegółów w pliku .pdf, .html lub .ipynb
    Parameters
    ----------
    x : tablica numpy.ndarray jednowymiarowa  
        (wektor)
    
    Returns
    -------
    Tablica numpy.ndarray jednowymiarowa  
    (wektor)
    """
    # 1.) np.cumsum() - skumulowane sumy
    w = np.cumsum(x)
    # 2.) odejmujemy od sum skumulowanych 
    #     wartości elementów tak by uzyskac tylko
    #     sumy k kolejnych elementów
    w[k:] -= w[:-k]
    # 3.) Wykorzystujemy wektoryzację wektor i skalar
    #     by każdą taka sumę podzielić przez k
    return w[(k-1):]/k

def srednia_ruchoma2(x, k):
    
    return np.convolve(x, np.ones(k), 'valid') / k

def srednia_ruchoma3(x, k):
    
    import pandas as pd

    w = pd.Series(x)

    return w.rolling(k).mean()


# Jako przykład wygenerujemy rysunek:
# Generujemy dane:
x = np.random.randn(100)
y = x.cumsum()
# Funkcja plt.plot() rysuje krzywą łącząc punkty zawarte
# w tablicy y
plt.plot(y, 'k-', label="dane") # 'k-' wygląd krzywej (czarna, ciągła)
k = 5
plt.plot(range(k-1, len(y)), srednia_ruchoma(y, k), 'r--', label="MA(3)") # 'r--` (czerwona, przerywana)
k = 10
plt.plot(range(k-1, len(y)), srednia_ruchoma(y, k), 'b-.', label="MA(5)") # 'b-.' (niebieska, kropka-kreska

plt.legend() # legenda (korzysta z etykiet label)
plt.show()

## Zadanie 1.12[MG]

Niech `t` będzie wektorem o $n$ elementach będących liczbami całkowitymi ze zbioru $\{1,\dots,k\}$. 
Napisz funkcję, która dokona kodowania elementów $t_i$ (*one-hot-encode*).
Funkcja powinna zwracać macierz zero-jedynkową $R$ wymiaru $n\times k$ taką, że 
$r_{i,j}=1$ wtedy i tylko wtedy gdy $t_i = j$. Taka reprezentacja jest przydatna 
np. w problemie klasyfikacji wieloetykietowej przy użyciu $k$ klasyfikatorów.


In [ ]:
import numpy as np

def onehotencode(t):
    """
    One-hot-encoder
    
    Parameters
    ----------
    t : vector (one dimensional numpy.ndarray array) 

    Returns
    -------
    Function return a 0-1 matrix of size nxk (numpy.ndarray)
    
    """
    assert t.ndim == 1
    
    n = t.shape[0] # liczba wierszy
    k = t.max()    # liczba kolumn wyniku
    # przygotujemy macierz odpowiedniego wymiaru
    # wypelnioną 0
    M = np.zeros((n, k)) 
    # Wykorzystamy indeksowanie: ciąg indeksów (lista) 
    # oraz ciąg indeksów (tablca t)
    # wybieramy elementy:
    # M[ind1[0], ind2[0]], M[ind1[1], ind2[1]], ....
    # i zmienimy ich wartość
    M[ [i for i in range(n)], t-1 ] = 1 # t-1 bo indeksujemy od 0
    # M[ list(range(n)), t-1] = 1 
    return(M)

def onehotencoder2(t):

    import pandas as pd
    #from sklearn.preprocessing import LabelEncoder
    from sklearn.preprocessing import OneHotEncoder
    
    # binary encode
    onehot_encoder = OneHotEncoder()
    return onehot_encoder.fit_transform(pd.DataFrame(t)).toarray()

def onehotedecode(X):
    """
    One-hot-decoder
    
    Parameters
    ----------
    X : matrix (two dimensional numpy.ndarray array) 

    Returns
    -------
    Function return a vector (one dimensional numpy.ndarray)
    sutch that for each row the number of the column with the
    value most close to 1 is returned
    """        
    # w której kolumnie wartość największa 
    # argmax zwraca indeks pierwszego maksimum (idziemy po kolumnach)
    return X.argmax(axis = 1) + 1

In [ ]:
t = np.array([1, 3, 1, 2, 5, 5])
print(onehotencode(np.array([1, 3, 1, 2, 5, 5])))
X = np.random.rand(12).reshape(-1, 3)


## Zadanie 1.13[MG]

Dokonaj przekształcenia \textit{softmax} każdego wiersza macierzy 
$\mathbf{X}\in\mathbb{R}^{n\times k}$, tzn. przekształcenia postaci: 
$$x_{i,j}\mapsto \frac{\exp(x_{i,j})}{\sum_{l=1}^k \exp(x_{i,l})}.$$

Następnie dokonaj odkodowania każdego wiersza (*one-hot decode*), 
tj. dla każdego wiersza należy znaleźć numer kolumny o wartości najbardziej 
zbliżonej do $1$. Zwróć wektor $n$-elementowy.


In [ ]:
def softmax1(X):
    r"""
    Softmax transformation
    i.e.
    x_{ij} -> ( exp(x_{ij}) / \sum_{l=1}^{k} exp_{x_{il}} )
    
    Parameters
    ----------
    X : matrix (two dimensional numpy.ndarray array) 

    Returns
    -------
    Function return transformed matrix of size nxk (numpy.ndarray)
    """    
    import scipy.special 
    
    return scipy.special.softmax(X, axis=1)

In [ ]:
def softmax2(X):
    r"""
    Softmax transformation
    i.e.
    x_{ij} -> ( exp(x_{ij}) / \sum_{l=1}^{k} exp_{x_{il}} )
    
    Parameters
    ----------
    X : matrix (two dimensional numpy.ndarray array) 

    Returns
    -------
    Function return transformed matrix of size nxk (numpy.ndarray)
    """    
    Y = np.exp(X) # funkcja zwektoryzowana
    # 1.) axis - po której będziemy liczyć sumę
    # tu 1 czyli dostaniemy wartości sumy w każdym wierszu
    # liczymy po kolumnach
    # 2.) (-1, 1) automatyczne dostosowanie liczby wierszy, 1 kolumna
    # dzieki temu dostaniemy wektoryzację po kolumnach
    return Y / Y.sum(axis = 1).reshape(-1, 1) 

In [ ]:
Y1 = softmax1(X)
Y2 = softmax2(X)

print(onehotedecode(Y1))
# itd.

onehotencoder2(np.array([1, 3, 1, 2, 5, 5]))

In [ ]:
Y3 = X*1000


In [ ]:
softmax1(Y3)

In [ ]:
softmax2(Y3)

In [ ]:
Y4 = Y3 - Y3.max(axis=1).reshape(-1, 1)
softmax2(Y4)